### ISSUE TITLE
Fail gracefully to locally installed Node.js version when nodejs.org is unreachable, and fail the build immediately if there is no applicable version locally

### ISSUE THREAD

**Issue by BanzaiMan** [travis-ci/travis-ci#5333]
When nodejs.org is unreachable for any reason, we fall back to `DEFAULT_VERSION`, which points to `0.10` at the moment. This default version may be too old (`0.10` definitely is right now; but in general, the discrepancy may be arbitrarily large or small).

Even when a usable version is locally available, it would not be used. For example, suppose we have

```yaml
node_js:
  - 4.2
```
and 4.2.0 pre-installed locally. If nodejs.org is unreachable, we fall back to 0.10 and continues the build. This is not ideal.

So, I propose that, for a given Node.js version, we should:

1. Install the latest version known to `nvm` if nodejs.org is reachable
1. If nodejs.org is not reachable, use the locally available and applicable version
1. If there is no version available locally, fail the build immediately

There is an alternative choice for the last scenario, that is, to continue with `DEFAULT_VERSION`, but it is quite likely that this is **not** what the user wants, since it is not running on the specified version.

**Comment by torra** [comment]
We recently had an issue because of this. Our travis build produces a single, deployable, ready-to-run artifact. We run node 4 in production, and when our build silently fell back to node 0.10 and succeeded, we deployed that code to production only to have it fail.

I would say it a little more strongly - falling back to some default version of node never makes sense if it does not match what is specified in `.travis.yml` (or `package.json`), because dependencies installed/compiled with node 0.10 are not compatible with node 4.

I think a good fix for this would be to simply fail the build when the required version of node is not available. Perhaps an optional flag in `.travis.yml` to allow falling back to `DEFAULT_VERSION` would be useful to some as well, but to me that is a nice-to-have feature and should certainly not be the default behavior.

**Comment by xaka** [comment]
+1 on failing the build immediately. Our just got broken due to failing back to node.js version we no longer support:

```
288.19s$ nvm install 4.2.6
#####################                                                     30.1%
curl: (18) transfer closed with 8272565 bytes remaining to read
Binary download failed, trying source.
$ node --version
v0.10.36
```
